### PR TITLE
fix(INF-1133): compare semantic single-block versions

### PR DESCRIPTION
**Pull request by Byron-X-22**
## Summary

- canonicalize embedded Solana JSON before computing the repair/retirement block digest, so field order and insignificant whitespace do not create false mismatches
- inspect and identity-check every immutable single-block version before pinning the current version; true payload differences, malformed JSON, delete markers, mutable/incomplete versions, and excessive version counts still fail closed
- extend the Docker lifecycle test to reproduce two semantically identical versions with different ETags, compressed sizes, and raw serialization

## Production evidence

`workflow.batch_consolidator/repair_existing_cscb_431572000_431872000_p20_r1_20260718T2330Z` failed after 225/300 CSCBs because 15 single-block keys in `[431632000,431647000)` each had two versions with different ETags and compressed sizes. Read-only inspection of all 30 immutable versions showed identical block identity, Chainstorage metadata, transaction metadata, and parsed Solana JSON for every pair; only JSON serialization differed.

No S3 or production database mutation is performed by this change. Existing pending repair manifests remain fenced for a deployment-time retry.

## Verification

- `env TEST_TYPE=unit go test ./...`
- `go test -race ./internal/storage/retirement ./internal/storage/cscbrepair`
- `go vet ./internal/storage/retirement ./internal/storage/cscbrepair`
- Docker-backed `TestIntegrationCSCBRepairFullLifecycle` with different-ETag/different-size semantic duplicate versions
- `git diff --check`

Linear: https://linear.app/cipherowl/issue/INF-1133

### CODE REVIEW

File: internal/storage/retirement/payload_verifier.go [hunk 1]
```diff
@@ -127,14 +129,46 @@ func (v *payloadVerifier) VerifyConsolidated(ctx context.Context, candidate Cand
...
+		}
+		return nil, xerrors.Errorf("failed to finish parsing embedded Solana block JSON: %w", err)
+	}
+	canonicalHeader, err := json.Marshal(value)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to canonicalize embedded Solana block JSON: %w", err)
+	}
+	solana.Header = canonicalHeader
+	return canonical, nil
+}
```
[the remark this answers is not in the document]
Comment (Byron-X-22) [reply internal/storage/retirement/payload_verifier.go:193]: Addressed in 6134488. The function now checks the original block for a Solana header first and clones only when it needs to replace the canonicalized JSON. Focused repair and retirement tests pass.

File: internal/storage/retirement/payload_verifier.go [hunk 2]
```diff
@@ -127,14 +129,46 @@ func (v *payloadVerifier) VerifyConsolidated(ctx context.Context, candidate Cand
 }
 
 func canonicalBlockDigest(block *api.Block) (string, error) {
-	payload, err := (proto.MarshalOptions{Deterministic: true}).Marshal(block)
+	canonical, err := canonicalizeEmbeddedBlockPayload(block)
```
Comment (watch-owl) [inline internal/storage/retirement/payload_verifier.go:156]: **Critical — version or bridge the persisted digest before changing its algorithm.** `PayloadSHA256` is durable workflow state in both retirement and CSCB repair manifests, but this changes it from deterministic protobuf over the raw Solana header to deterministic protobuf over canonicalized JSON. Any pre-deploy manifest whose header was not already in `encoding/json` canonical order will therefore recompute a different digest even though the pinned object bytes and block semantics are unchanged. Retirement resume rejects that at `planner.go:516/523`; repair resume rejects it at `repairer.go:301/376` (and a `StatePrepared` repair can first pass `Restore`, mutate metadata, then stall in `VerifyRebuilt`). A rollback after creating canonical-digest manifests has the inverse failure. Please preserve upgrade/rollback compatibility—for example with an explicit digest version/migration or dual legacy+canonical verification—and add a resume test that crosses the old/new algorithm boundary.

**Review by watch-owl — CHANGES_REQUESTED** [review]
Requesting changes for persisted-digest upgrade/rollback compatibility. The semantic duplicate comparison and fail-closed topology checks otherwise look sound, and CircleCI is green at the reviewed head.

File: internal/storage/retirement/payload_verifier.go [hunk 3]
```diff
@@ -127,14 +129,46 @@ func (v *payloadVerifier) VerifyConsolidated(ctx context.Context, candidate Cand
...
 	return hex.EncodeToString(digest[:]), nil
 }
 
+func canonicalizeEmbeddedBlockPayload(block *api.Block) (*api.Block, error) {
+	solana := block.GetSolana()
+	if solana == nil || len(solana.Header) == 0 {
+		return block, nil
+	}
+
+	decoder := json.NewDecoder(bytes.NewReader(solana.Header))
```
Comment (watch-owl) [inline internal/storage/retirement/payload_verifier.go:174]: **Warning — reject duplicate object keys before treating JSON as semantically identical.** Decoding into `any` stores objects in a Go map, so `encoding/json` silently overwrites duplicate members. For example, this exact decode/marshal path maps both `{"slot":1,"slot":2}` and `{"slot":2}` to `{"slot":2}`. Those raw headers are different and the first is ambiguous across JSON consumers, yet the verifier would accept them as the same payload before retiring the single-block copy. That violates the stated fail-closed behavior for true/malformed payload differences. Please detect and reject duplicate keys at every object depth (or use a strict canonicalizer that does) and add top-level and nested regression tests.

File: internal/storage/retirement/payload_verifier.go (hunk 2 above)
Comment (Byron-X-22) [reply internal/storage/retirement/payload_verifier.go:156]: Fixed in `863f122`. `PayloadSHA256` remains the existing deterministic-protobuf digest over the storage-placement-neutral raw block, so pre-deploy manifests resume unchanged and manifests created by this code remain readable after rollback. A separate `SemanticSHA256` is computed only for semantic duplicate-version and source-CSCB comparison and is never persisted. Added `TestVerifyRebuiltResumesManifestWithPreSemanticSolanaDigest`, which resumes a `StateRestored` manifest carrying the pre-change digest over deliberately non-canonical Solana JSON and completes payload verification. Focused tests, race tests, vet, the full unit suite, and the Docker-backed PostgreSQL and LocalStack lifecycle test pass.

File: internal/storage/retirement/payload_verifier.go (hunk 3 above)
Comment (Byron-X-22) [reply internal/storage/retirement/payload_verifier.go:174]: Fixed in `f7c6d71`. Canonicalization now uses a recursive token parser that rejects duplicate decoded object keys at every nesting depth before building any map. Added top-level and nested duplicate-key regression tests; distinct valid values remain significant. Focused tests, race tests, vet, the full unit suite, and the Docker-backed PostgreSQL and LocalStack lifecycle test pass.

**Review by watch-owl — APPROVED** [review]
Re-reviewed exact head `f7c6d71694e782a42645dbeca712b3c11ace0d1a`; prior blockers are resolved. Persisted `PayloadSHA256` keeps the legacy deterministic-protobuf algorithm while the non-persisted semantic digest is used only for semantic comparisons, and duplicate JSON keys are rejected recursively. Focused unit/race tests and vet passed in Docker, CircleCI is green, and I found no remaining blocking issues.
